Formula/spago.rb

```diff
@@ -0,0 +1,23 @@
+class Spago < Formula
+  desc "PureScript package manager and build tool"
+  homepage "https://github.com/purescript/spago"
+  url "https://github.com/purescript/spago/archive/refs/tags/0.20.4.tar.gz"
+  sha256 "e2ef8604115556b39ec71301d85a3502502fd5972e40bac82cd556d6a128baff"
+  license "BSD-3-Clause"
+  head "https://github.com/purescript/spago.git", branch: "master"
+
+  depends_on "ghc" => :build
+  depends_on "haskell-stack" => :build
+  depends_on "purescript"
+
+  def install
+    system "make", "fetch-templates"
```

Comment:
Can we declare these templates as `resource`s instead? That way they're properly cached and checksummed.

Reply:
It invokes [this shell script](https://github.com/purescript/spago/blob/master/scripts/fetch-templates) under the hood.  It downloads 4 files and put them in templates/ directory.  I tried to do the same thing with `resource` instead of the shell script, but apparently something went wrong; fetched files are not placed:

```ruby
  resource "docs-search-app-0.0.10.js" do
    url "https://github.com/purescript/purescript-docs-search/releases/download/v0.0.10/docs-search-app.js"
    sha256 "45dd227a2139e965bedc33417a895ec7cb267ae4a2c314e6071924d19380aa54"
  end

  resource "docs-search-app-0.0.11.js" do
    url "https://github.com/purescript/purescript-docs-search/releases/download/v0.0.11/docs-search-app.js"
    sha256 "0254c9bd09924352f1571642bf0da588aa9bdb1f343f16d464263dd79b7e169f"
  end

  resource "purescript-docs-search-0.0.10" do
    url "https://github.com/purescript/purescript-docs-search/releases/download/v0.0.10/purescript-docs-search"
    sha256 "437ac8b15cf12c4f584736a07560ffd13f4440cd0c44c3a6f7a29248a1ff8171"
  end

  resource "purescript-docs-search-0.0.11" do
    url "https://github.com/purescript/purescript-docs-search/releases/download/v0.0.11/purescript-docs-search"
    sha256 "06dfcb9b84408527a2980802108fae6a5260a522013f67d0ef7e83946abe4dc2"
  end

  def install
    # Equivalent to make fetch-templates:
    resource("docs-search-app-0.0.10.js").stage buildpath/"templates"
    resource("docs-search-app-0.0.11.js").stage buildpath/"templates"
    resource("purescript-docs-search-0.0.10").stage buildpath/"templates"
    resource("purescript-docs-search-0.0.11").stage buildpath/"templates"

    system "stack", "install", "--system-ghc", "--no-install-ghc", "--skip-ghc-check", "--local-bin-path=#{bin}"
    (bash_completion/"spago").write `#{bin}/spago --bash-completion-script #{bin}/spago`
    (zsh_completion/"_spago").write `#{bin}/spago --zsh-completion-script #{bin}/spago`
  end
```

```
$ brew reinstall --build-from-source spago
==> Downloading https://github.com/purescript/purescript-docs-search/releases/download/v0.0.10/docs-search-app.js
Already downloaded: /Users/dahlia/Library/Caches/Homebrew/downloads/3e4f563172cf1987c61c707b1636b4f4f8bb647b1dae68cac7bb5c4642a66018--docs-search-app.js
==> Downloading https://github.com/purescript/purescript-docs-search/releases/download/v0.0.11/docs-search-app.js
Already downloaded: /Users/dahlia/Library/Caches/Homebrew/downloads/93f148a4d84ff51bb90f65be036ea334d7af4093cbb974e16faeb5ac418ca573--docs-search-app.js
==> Downloading https://github.com/purescript/purescript-docs-search/releases/download/v0.0.10/purescript-docs-search
Already downloaded: /Users/dahlia/Library/Caches/Homebrew/downloads/f023cff93e5e1b388310b77a377f465b9e5b2b17cc50f83071728db8dbf19223--purescript-docs-search
==> Downloading https://github.com/purescript/purescript-docs-search/releases/download/v0.0.11/purescript-docs-search
Already downloaded: /Users/dahlia/Library/Caches/Homebrew/downloads/5772dc67cbf7ef42b8fe12a7a3acf46990a150e8bf2454b6e138994442365ac5--purescript-docs-search
==> Downloading https://github.com/purescript/spago/archive/refs/tags/0.20.4.tar.gz
Already downloaded: /Users/dahlia/Library/Caches/Homebrew/downloads/80d269ba019ea890d97e8f38e981165263030031ddd180cb0034eeb5e8f86fa9--spago-0.20.4.tar.gz
==> Reinstalling spago 
==> stack install --system-ghc --no-install-ghc --skip-ghc-check --local-bin-path=/opt/homebrew/Cellar/spago/0.20.4/bin
Last 15 lines from /Users/dahlia/Library/Logs/Homebrew/spago/01.stack:
30 | docsSearchApp0010 = $(embedFileUtf8 "templates/docs-search-app-0.0.10.js")
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Warning: File listed in spago.cabal file does not exist: templates/docs-search-app-0.0.10.js

Warning: File listed in spago.cabal file does not exist: templates/docs-search-app-0.0.11.js

Warning: File listed in spago.cabal file does not exist: templates/purescript-docs-search-0.0.10

Warning: File listed in spago.cabal file does not exist: templates/purescript-docs-search-0.0.11
Completed 167 action(s).

--  While building package spago-0.20.4 (scroll up to its section to see the error) using:
      /private/tmp/spago-20220208-88758-bvl833/spago-0.20.4/.brew_home/.stack/setup-exe-cache/aarch64-osx/Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.7 --builddir=.stack-work/dist/aarch64-osx/Cabal-3.2.1.0 build lib:spago exe:spago --ghc-options ""
    Process exited with code: ExitFailure 1
```

Could you let me know what I did wrong?:

Reply:
Try doing `brew install --keep-temp spago` and check where `brew` put the resources.

My guess is that they just have the wrong extensions.

Reply:
Actually, I think this should work
```ruby
(buildpath/"templates").install resources
```

Reply:
`(buildpath/"templates").install resources` didn't work, as [`Pathname#install`][1] does not directly take an array.  Instead, I tried to expand `resources` before passing it: `(buildpath/"templates").install *resources`.  Unfortunately, it either didn't work, because there are two pairs of files with the same name.  They should be renamed to names with version-suffixed.  My last attempt worked and looks like below, but it's not that clean:

```ruby
# Equivalent to make fetch-templates:
(buildpath/"templates").cd do |d|
  d.install resource("docs-search-app-0.0.10.js"), resource("purescript-docs-search-0.0.10")
  (d/"docs-search-app.js").rename "docs-search-app-0.0.10.js"
  (d/"purescript-docs-search").rename "purescript-docs-search-0.0.10"
  d.install resource("docs-search-app-0.0.11.js"), resource("purescript-docs-search-0.0.11")
  (d/"docs-search-app.js").rename "docs-search-app-0.0.11.js"
  (d/"purescript-docs-search").rename "purescript-docs-search-0.0.11"
end
```

If you let me know how to make it better I'm going to adjust it.  Thanks!


[1]: https://rubydoc.brew.sh/Pathname.html#install-instance_method